File: solutions/python/dnd-character/1/dnd_character.py
from random import randint


class Character:
    def __init__(self):
        self._strength = self.ability()
        self._dexterity = self.ability()
        self._constitution = self.ability()
        self._intelligence = self.ability()
        self._wisdom = self.ability()
        self._charisma = self.ability()
        self.hitpoints = 10 + modifier(self.constitution)

    def ability(self):
        return roll_ability()

    @property
    def strength(self):
        return self._strength

    @property
    def dexterity(self):
        return self._dexterity

    @property
    def constitution(self):
        return self._constitution

    @property
    def intelligence(self):
        return self._intelligence

    @property
    def wisdom(self):
        return self._wisdom

    @property
    def charisma(self):
        return self._charisma


def modifier(value):
    return (value - 10) // 2


def roll_ability():
    dices = [randint(1, 6) for _ in range(4)]
    return sum(dices) - min(dices)

File: solutions/cpp/dnd-character/1/dnd_character.cpp
#include "dnd_character.h"

#include <algorithm>
#include <random>
#include <vector>

namespace dnd_character {
    int modifier(const int constitution) {
        const int diff = constitution - 10;
        int mod = diff / 2;
        if (diff < 0 && diff % 2 != 0) {
            mod--;
        }
        return mod;
    }

    int ability() {
        static std::random_device rd;
        static std::mt19937 gen(rd());
        static std::uniform_int_distribution<> dist(1, 6);

        std::vector<int> dices;

        for (int i = 0; i < 4; ++i) {
            dices.push_back(dist(gen));
        }

        std::sort(dices.begin(), dices.end());

        return dices[1] + dices[2] + dices[3];
    }
} // namespace dnd_character
